core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
One concern here is that in relinquishing the controller lock above, we could see `IllegalStateException`s due to other triggers to the `PartitionStateMachine`. E.g., if a broker comes up or goes down, it would invoke `PartitionStateMachine.handleStateChanges` which expects an empty broker request batch. Since it is non-empty at this point it will throw the ISE.